### PR TITLE
Ssh result error handling

### PR DESCRIPTION
**Pull request by hustodemon**

### CODE REVIEW

File: src/main/java/com/suse/salt/netapi/parser/ResultSSHResultTypeAdapterFactory.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+                                .getAsJsonPrimitive().isString()) {
+                    return Optional.of(json.getAsJsonObject().get("stderr")
+                            .getAsJsonPrimitive().getAsString());
+                }
+
+                return Optional.empty();
+            }
+        };
+    }
+
```
Comment (moio) [inline src/main/java/com/suse/salt/netapi/parser/ResultSSHResultTypeAdapterFactory.java]: Too many blank lines here I guess.

File: src/test/resources/salt_ssh_output_success_empty_return.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
+    "fun": "test.ping",
+    "fun_args": ["arg1", "arg2"],
+    "id": "silver",
+    "jid": "20141202152721523072",
+    "retcode": 0,
+    "return": null
+}
```
Comment (moio) [inline src/test/resources/salt_ssh_output_success_empty_return.json]: Missing newline here.

File: src/main/java/com/suse/salt/netapi/parser/ResultSSHResultTypeAdapterFactory.java (hunk 1 above)
Comment (isbm) [inline src/main/java/com/suse/salt/netapi/parser/ResultSSHResultTypeAdapterFactory.java]:
```
zypper in auto-moio-syntax-checker
```

_/me runs away_ :laughing:

File: src/test/java/com/suse/salt/netapi/parser/JsonParserTest.java [hunk 3]
```diff
@@ -304,4 +310,75 @@ public void testJobsMultipleKwargs() throws Exception {
...
+        InputStream is = this.getClass()
+                .getResourceAsStream("/salt_ssh_output_success_nonzero_return_code.json");
+        JsonParser<Result<SSHResult<String>>> resultJsonParser =
+                new JsonParser<>(new TypeToken<Result<SSHResult<String>>>() {});
+
+        Result<SSHResult<String>> result = resultJsonParser.parse(is);
+        assertFalse(result.error().isPresent());
+        assertTrue(result.result().isPresent());
+        SSHResult<String> sshResult = result.result().get();
+        assertEquals(true, sshResult.getReturn().isPresent());
```
Comment (lucidd) [inline src/test/java/com/suse/salt/netapi/parser/JsonParserTest.java]: assertTrue

File: src/test/java/com/suse/salt/netapi/parser/JsonParserTest.java [hunk 4]
```diff
@@ -304,4 +310,75 @@ public void testJobsMultipleKwargs() throws Exception {
...
+        InputStream is = this.getClass()
+                .getResourceAsStream("/salt_ssh_output_success_empty_return.json");
+        JsonParser<Result<SSHResult<Boolean>>> resultJsonParser =
+                new JsonParser<>(new TypeToken<Result<SSHResult<Boolean>>>() {});
+
+        Result<SSHResult<Boolean>> result = resultJsonParser.parse(is);
+        assertFalse(result.error().isPresent());
+        assertTrue(result.result().isPresent());
+        SSHResult<Boolean> sshResult = result.result().get();
+        assertEquals(false, sshResult.getReturn().isPresent());
```
Comment (lucidd) [inline src/test/java/com/suse/salt/netapi/parser/JsonParserTest.java]: assertFalse

File: src/main/java/com/suse/salt/netapi/utils/SaltErrorUtils.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+            Pattern.compile("'([^']+)' is not available.");
+    private static final Pattern MODULE_NOT_SUPPORTED =
+            Pattern.compile("'([^']+)' __virtual__ returned False");
+
+    /**
+     * Based on the salt text output, derive particular {@link SaltError}.
+     * @param saltOutput salt output
+     * @return salt error corresponding to given string
+     */
+    public static Optional<SaltError> deriveError(Optional<String> saltOutput) {
```
Comment (lucidd) [inline src/main/java/com/suse/salt/netapi/utils/SaltErrorUtils.java]: the input does not need to be optional here

Comment (hustodemon) [inline src/main/java/com/suse/salt/netapi/utils/SaltErrorUtils.java]: I made it optional to be able to control flow as this method is fed from the `extractStdErr`, which returns optional. Alternative would be to use null, or to introduce some code branching in the callers of `deriveError`. What do you think?

Comment (lucidd) [inline src/main/java/com/suse/salt/netapi/utils/SaltErrorUtils.java]: flamap

Comment (hustodemon) [inline src/main/java/com/suse/salt/netapi/utils/SaltErrorUtils.java]: Nice picture, but it'd be faster for all if you provided a constructive feedback (code snippet would suffice).

Comment (lucidd) [inline src/main/java/com/suse/salt/netapi/utils/SaltErrorUtils.java]: `extractStdErr(json).flatMap(deriveError)`

Comment (isbm) [inline src/main/java/com/suse/salt/netapi/utils/SaltErrorUtils.java]: LOL

**Comment by hustodemon** [comment]
Addressed the issues in the latest commits.

**Comment by hustodemon** [comment]
Guys, can we merge the PR?

**Comment by lucidd** [comment]
threre is one more `assertEquals(true` after that should be good to merge.

**Comment by hustodemon** [comment]
Oh, it escaped my attention, thanks :) Force pushed the fix in 6fdfc78.

**Pull request merged**